### DISCUSSION TITLE
Multiple Responses Models

### DISCUSSION THREAD

**Discussion by AndreMPCosta** [fastapi/fastapi#7109] [Questions] [answered]
### Example

**Resource:**
```Python
@router.get("/season/{season_id}", response_model=Union[Split, Season], response_model_exclude={'id'})
async def get_season(season_id: int, split: Optional[int] = None):
    season = await Season.get_by_season_id(season_id)
    if not season:
        return JSONResponse(status_code=404, content={"message": "Season not found"})
    if split:
        return await Split.get_by_number(season.id, split)
    else:
        return season
```

**Models**
* Season:
```Python
class Season(Model, ABC):
    season_id: Optional[int] = Field(1, ge=1)
    start_date: Optional[datetime] = datetime(date.today().year, 1, 1)
    end_date: Optional[datetime] = datetime(date.today().year, 12, 31)
    splits: Optional[List[ObjectId]] = []

    async def create(self) -> None:
        self.season_id = await Season.count_seasons() + 1
        await self.save()

    @classmethod
    async def get_by_id(cls, _id: int) -> "Season":
        return await engine.find_one(cls, cls.id == _id)
```

* Split:
```Python
class Split(Model, ABC):
    split_id: Optional[int] = Field(1, ge=1)
    number: Optional[int] = Field(1, ge=1)
    start_date: Optional[datetime] = datetime(date.today().year, date.today().month, 1)
    end_date: Optional[datetime]
    season: Optional[ObjectId]

    async def create(self) -> None:
        season = await Season.get_last()
        self.split_id = await self.count_splits() + 1
        self.number = season.count_splits() + 1
        self.start_date = datetime(date.today().year, date.today().month, 1)
        self.end_date = self.last_day_of_month(self.start_date)
        self.season = season.id
        await self.save()
        season.splits.append(self.id)
        await season.save()

    @classmethod
    async def get_by_number(cls, season_id, split_number: int) -> "Split":
        return await engine.find_one(cls, cls.season == season_id, cls.number == split_number)
```

### Description

* I expect to return a season model when the query is empty, and a split model if the query is not empty
* When I try to get "/season/3?split=2" it returns
```JSON
{
    "split_id": 2,
    "number": 2,
    "start_date": "2020-12-01T00:00:00",
    "end_date": "2020-12-31T00:00:00",
    "season": "5fd894151470b8ec917c6007"
}
```
as expected.
* But when I try to get "/season/3" it returns
```JSON
{
    "split_id": 1,
    "number": 1,
    "start_date": "2020-01-01T00:00:00",
    "end_date": "2020-12-31T00:00:00",
    "season": null
}
```
which is still a Split Model and not a Season Model, even though the method returns a Season Model and not a Split Model.

The expected output would be:
```JSON
{
    "season_id": 3,
    "start_date": "2020-01-01T00:00:00",
    "end_date": "2020-12-31T00:00:00",
    "splits": [
        "5fd894271470b8ec917c6008",
        "5fd8942e1470b8ec917c6009"
    ]
}
```
which I achieve if I swap Split and Season on:
```Python
@router.get("/season/{season_id}", response_model=Union[Season, Split], response_model_exclude={'id'})
```

So it seems that only the first model is being used.

### Environment

* OS: Windows 10
* FastAPI Version: 0.62.0

* Python version: 3.9.1

**Comment by ycd** [comment]
There must be something wrong with your function. `response_model`'s only purpose is validating your return value and documenting it on OpenAPI schema. It doesn't change the way of how your coroutine/function works.

**Comment by AndreMPCosta** [comment]
How is something wrong with my function, if I showed you if I swapped the order of the response models, the model that was not showing correctly starts to show correctly? (Of course the other model won't work, only the **first** model on the Union works).

Like I literally tried this on Postman two minutes ago, if my coroutines were wrong, swapping models on Union wouldn't change the outcome.

Unless this is not the way of doing what I want, which is to pick an appropriate response based on the return Model.

**Comment by Mause** [comment]
@AndreMPCosta please read the documentation regarding how pydantic chooses which model in a Union to use: https://pydantic-docs.helpmanual.io/usage/types/#unions

**Comment by AndreMPCosta** [comment]
@Mause Thanks for pointing me to that.
However after reading I encounter this bit:

>  pydantic will attempt to 'match' any of the types defined under Union and will use the first one that matches

> As such, it is recommended that, when defining Union annotations, the most specific type is included first and followed by less specific types

Since both are a odmantic model, I'm kind of screwed, it will try to "convert" the season Model to Split, or vice-versa. 
Any ideas, how to manage to do what I have stated before?

For now I'll lose documentation, it works without the response_model on the decorator.

In theory it should work reading the docs: 
https://fastapi.tiangolo.com/tutorial/extra-models/?h=+dict#union-or-anyof

**Comment by ycd** [comment]
@AndreMPCosta sorry, i completely missed the part that it works when you reverse the order of models in `response_model`.

**Comment by AndreMPCosta** [comment]
> @AndreMPCosta sorry, i completely missed the part that it works when you reverse the order of models in `response_model`.

Yeah I wasn't understanding your response, no worries, we all read something, somewhere, in a rush!

**Comment by hellocoldworld** [comment]
I have run into these a couple of times. My ugly workaround is to do:

```
from enum import Enum

class SeasonType(str, Enum):
    season = "season"

class SplitType(str,  Enum):
    split = "split"

class Season(BaseModel):
      model_type: SeasonType = Field(..., )
      ...

class Split(BaseModel):
     model_type: SplitType = Field(..., )
     ....
```

And sending in the response which model I'm using. It's ugly, but gets the job done. Curious of hearing other ideas, though.

**Comment by Mause** [comment, marked as answer]
@hellocoldworld I think you'll find that's exactly why `Literal` exists

```python

import json
from pydantic import BaseModel
from typing import Literal, Union

class Season(BaseModel):
    model_type: Literal['season']
    hello: int

class Split(BaseModel):
    model_type: Literal['split']
    hello: str


class Parent(BaseModel):
    thing: Union[Season, Split]
```

**Comment by pythonweb2** [comment]
Constructing the pydantic model object prior to returning it ensures that the data is valid, so you can directly return a JSONResponse after using the jsonable_encoder on the pydantic object.

Keep in mind that you will still want to keep the type reference in the `response_model` parameter so that your OpenAPI docs will still work.

References:
- https://fastapi.tiangolo.com/advanced/custom-response/
- https://fastapi.tiangolo.com/tutorial/encoder/?h=jsonable_en#using-the-jsonable_encoder

**Comment by fabioueno** [comment]
Hey there, @AndreMPCosta!

Is your issue resolved? 🤔

**Comment by AndreMPCosta** [comment]
Hey,

I've done a workaround during the time I've opened the issue, still don't know if Union of multiple types is working or not. You can close the issue, I guess.
